### PR TITLE
Array binary search

### PR DESCRIPTION
**Pull request by JamesDansie**
- [ ] Top-level README “Table of Contents” is updated
 - [ x] Feature tasks for this challenge are completed
 - [ x] Unit tests written and passing
     - [x ] “Happy Path” - Expected outcome
     - [ x] Expected failure
     - [ x] Edge Case (if applicable/obvious)
 - [ x] README for this challenge is complete
     - [x ] Summary, Description, Approach & Efficiency, Solution
     - [ x] Link to code
     - [ x] Picture of whiteboard

### CODE REVIEW

File: code401challenges/src/test/java/code401challenges/BinarySearchTest.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+                "This should not find anything",
+                -1,
+                BinarySearch.binarySearch(new int[]{1, 2, 3, 4}, 5)
+                );
+    }
+
+    @Test
+    public void testSmall() {
+        assertEquals(
+                "This should not find anything",
```
Comment (ncarignan) [inline code401challenges/src/test/java/code401challenges/BinarySearchTest.java]: this one should find 2

File: code401challenges/src/test/java/code401challenges/BinarySearchTest.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+                1,
+                BinarySearch.binarySearch(new int[]{1, 2, 3, 4, 7, 9, 10, 20, 30}, 2)
+        );
+    }
+
+    @Test
+    public void testBig() {
+        assertEquals(
+                "This should not find anything",
```
Comment (ncarignan) [inline code401challenges/src/test/java/code401challenges/BinarySearchTest.java]: this one should find 7

File: code401challenges/src/main/java/code401challenges/BinarySearch.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+public class BinarySearch {
+    //this is very helpful; https://en.wikipedia.org/wiki/Binary_search_algorithm
+    public static int binarySearch(int[] arrIn, int key){
+        int lower = 0;
+        int upper = arrIn.length - 1;
+        while(lower <= upper){
+            int mid = (lower + upper)/2;
+
+//            String out = String.format("Lower is %d Upper is %d Mid is %d", lower, upper, mid);
```
Comment (ncarignan) [inline code401challenges/src/main/java/code401challenges/BinarySearch.java]: I like that you are practicing String.format, I would remove this from the solution code though. We will be taking off points soon for "corpse code"
